File: backend/src/modules/module_manager/schemas.py
"""模块管理模块 - Pydantic Schema。"""
from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class MenuItem(BaseModel):
    id: str
    parent_id: Optional[str] = None
    title: str
    icon: Optional[str] = None
    path: str
    component: str
    permission: Optional[str] = None
    order: int = 0


class ModuleInstallReq(BaseModel):
    """从受控目录安装模块（source_path 必须在 MODULE_UPLOAD_DIR 下）。"""
    install_type: Literal["path"] = Field("path", description="v1.1 起仅支持 path（ZIP 请使用 /upload）")
    source_path: str = Field(..., description="源码目录路径（必须位于 MODULE_UPLOAD_DIR 下）")


class ModuleUpgradeReq(BaseModel):
    """升级模块。"""
    install_type: Literal["zip", "path"] = Field(..., description="升级源类型")
    file_path: Optional[str] = Field(None, description="ZIP 包路径（zip 时必填）")
    source_path: Optional[str] = Field(None, description="源码目录路径（path 时必填）")


class ModuleConfigUpdate(BaseModel):
    config: Dict[str, Any] = Field(..., description="模块配置对象")


class ModuleOut(BaseModel):
    id: str
    name: str
    version: str
    description: Optional[str] = None
    author: Optional[str] = None
    homepage: Optional[str] = None
    status: str
    entry_backend: str
    entry_frontend: Optional[str] = None
    dependencies: List[str] = []
    menus: List[MenuItem] = []
    config: Dict[str, Any] = {}
    installed_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ModuleUploadResp(BaseModel):
    """ZIP 上传安装响应。"""
    module_id: str
    version: str
    status: str
    saved_path: str

File: backend-go/internal/modules/module_manager/schemas.go
package module_manager
import "time"
// MenuItem 菜单项。
type MenuItem struct {
	ID         string  `json:"id"`
	ParentID   *string `json:"parent_id"`
	Title      string  `json:"title"`
	Icon       *string `json:"icon"`
	Path       string  `json:"path"`
	Component  string  `json:"component"`
	Permission *string `json:"permission"`
	Order      int     `json:"order"`
}
// ModuleInstallReq 从受控目录安装模块。
type ModuleInstallReq struct {
	InstallType string `json:"install_type" binding:"omitempty,oneof=path"`
	SourcePath  string `json:"source_path" binding:"required"`
}
// ModuleUpgradeReq 升级模块。
type ModuleUpgradeReq struct {
	InstallType string  `json:"install_type" binding:"required,oneof=zip path"`
	FilePath    *string `json:"file_path"`
	SourcePath  *string `json:"source_path"`
}
// ModuleConfigUpdate 更新模块配置。
type ModuleConfigUpdate struct {
	Config map[string]interface{} `json:"config" binding:"required"`
}
// ModuleOut 模块响应对象。
type ModuleOut struct {
	ID            string                 `json:"id"`
	Name          string                 `json:"name"`
	Version       string                 `json:"version"`
	Description   *string                `json:"description"`
	Author        *string                `json:"author"`
	Homepage      *string                `json:"homepage"`
	Status        string                 `json:"status"`
	EntryBackend  string                 `json:"entry_backend"`
	EntryFrontend *string                `json:"entry_frontend"`
	Dependencies  []string               `json:"dependencies"`
	Menus         []MenuItem             `json:"menus"`
	Config        map[string]interface{} `json:"config"`
	InstalledAt   string                 `json:"installed_at"`
	UpdatedAt     string                 `json:"updated_at"`
}
// ModuleListQuery 模块列表查询参数。
type ModuleListQuery struct {
	Page        int    `form:"page" binding:"omitempty,min=1"`
	PageSize    int    `form:"page_size" binding:"omitempty,min=1,max=100"`
	Status      string `form:"status" binding:"omitempty,oneof=active inactive"`
	Keyword     string `form:"keyword"`
	FilterMenus bool   `form:"filter_menus"`
}
// 时间辅助。
func formatDateTime(t time.Time) string {
	if t.IsZero() {
		return ""
	}
	t = t.UTC()
	if t.Nanosecond()/1000 == 0 {
		return t.Format("2006-01-02T15:04:05")
	}
	return t.Format("2006-01-02T15:04:05.000000")
}
